File: packages/sdk-ts/src/agents.ts
// suite.agents.* — invoke AgentField agents via the gateway.
//
// Maps 1:1 to TECH-SPEC.md §4.4. All model calls go through AF; this
// module never reaches a model provider directly.

import { z } from "zod"
import {
  parseSse,
  rawRequest,
  request,
  toCamel,
  type HttpOptions,
  type SseEvent,
} from "./_http.js"

// ---------- shared schemas ----------

const executionStatusSchema = z.enum([
  "queued",
  "running",
  "paused",
  "succeeded",
  "failed",
  "cancelled",
])

const callResultSchema = z.object({
  executionId: z.string(),
  status: executionStatusSchema,
  output: z.unknown().optional(),
  durationMs: z.number().optional(),
  error: z
    .object({
      code: z.string().optional(),
      message: z.string().optional(),
    })
    .optional(),
})
export type CallResult = z.infer<typeof callResultSchema>

const asyncCallResultSchema = z.object({
  executionId: z.string(),
  status: executionStatusSchema,
})
export type AsyncCallResult = z.infer<typeof asyncCallResultSchema>

const executionStatusResponseSchema = z.object({
  executionId: z.string(),
  status: executionStatusSchema,
  output: z.unknown().optional(),
  error: z
    .object({
      code: z.string().optional(),
      message: z.string().optional(),
    })
    .optional(),
  startedAt: z.string().optional(),
  finishedAt: z.string().optional(),
  durationMs: z.number().optional(),
})
export type ExecutionStatus = z.infer<typeof executionStatusResponseSchema>

const pendingApprovalSchema = z.object({
  executionId: z.string(),
  agent: z.string(),
  reason: z.string().optional(),
  requestedAt: z.string().optional(),
  payload: z.unknown().optional(),
})
export type PendingApproval = z.infer<typeof pendingApprovalSchema>

const pendingApprovalsResponseSchema = z.object({
  items: z.array(pendingApprovalSchema),
})

// ---------- input shapes ----------

export interface CallOptions extends HttpOptions {
  /** Server-side execution timeout in seconds (forwarded to AF). */
  timeoutS?: number
  /** Idempotency key — server dedupes within its retention window. */
  idempotencyKey?: string
}

export interface CallAsyncOptions extends CallOptions {
  /** Webhook URL fired when the async execution finishes. */
  webhookUrl?: string
}

export interface ApproveOptions extends HttpOptions {
  byUserId?: string
  note?: string
}

export interface DenyOptions extends HttpOptions {
  byUserId?: string
  reason?: string
}

// ---------- helpers ----------

function encodeAgent(name: string): string {
  // Agent names look like "ns.func". We pass them verbatim except for
  // characters that would break the URL path. AF's gateway accepts dots
  // and dashes natively; everything else gets percent-encoded.
  return encodeURIComponent(name).replace(/%2E/g, ".")
}

function buildCallBody(
  input: unknown,
  extras: Record<string, unknown>,
): Record<string, unknown> {
  // Agent input is forwarded verbatim — agent-defined schemas may carry
  // semantically-meaningful keys (e.g. `result_text` vs `resultText`)
  // and we must not silently rename them. Envelope fields below
  // (timeout_s, webhook_url, etc.) are SDK-defined and translated.
  const body: Record<string, unknown> = { input }
  for (const [k, v] of Object.entries(extras)) {
    if (v !== undefined) body[k] = v
  }
  return body
}

/**
 * Convert only the envelope fields of an execution response to camelCase,
 * leaving user-defined payload fields (`output`, `payload`, `input`)
 * untouched — those follow agent-defined schemas the SDK must not rename.
 */
const OPAQUE_FIELDS = ["output", "payload", "input"] as const

function camelEnvelope(raw: unknown): unknown {
  if (typeof raw !== "object" || raw === null || Array.isArray(raw)) return raw
  const source = raw as Record<string, unknown>
  const stripped: Record<string, unknown> = { ...source }
  const preserved: Record<string, unknown> = {}
  for (const field of OPAQUE_FIELDS) {
    if (field in source) {
      preserved[field] = source[field]
      delete stripped[field]
    }
  }
  const camelRest = toCamel<Record<string, unknown>>(stripped)
  return { ...camelRest, ...preserved }
}

// ---------- public API ----------

/** Sync agent invocation. Returns once AF has produced a result. */
export async function call(
  agent: string,
  input: unknown,
  opts: CallOptions = {},
): Promise<CallResult> {
  const { timeoutS, idempotencyKey, ...http } = opts
  const body = buildCallBody(input, {
    timeout_s: timeoutS,
    idempotency_key: idempotencyKey,
  })
  const raw = await request<unknown>("POST", `/agents/${encodeAgent(agent)}`, body, http)
  return callResultSchema.parse(camelEnvelope(raw))
}

/**
 * Fire-and-track agent invocation. Returns immediately with an execution
 * id; poll with `status()` or stream with `stream()` to observe progress.
 */
export async function callAsync(
  agent: string,
  input: unknown,
  opts: CallAsyncOptions = {},
): Promise<AsyncCallResult> {
  const { timeoutS, idempotencyKey, webhookUrl, ...http } = opts
  const body = buildCallBody(input, {
    timeout_s: timeoutS,
    idempotency_key: idempotencyKey,
    webhook_url: webhookUrl,
  })
  const raw = await request<unknown>(
    "POST",
    `/agents/async/${encodeAgent(agent)}`,
    body,
    http,
  )
  return asyncCallResultSchema.parse(camelEnvelope(raw))
}

/**
 * Stream agent events via SSE. Yields `SseEvent`s; the caller is
 * responsible for interpreting event names. Closes when the server
 * terminates the stream or the supplied AbortSignal fires.
 */
export async function* stream(
  agent: string,
  input: unknown,
  opts: CallOptions = {},
): AsyncIterable<SseEvent> {
  const { timeoutS, idempotencyKey, ...http } = opts
  const body = buildCallBody(input, {
    timeout_s: timeoutS,
    idempotency_key: idempotencyKey,
  })
  const headers = { ...(http.headers ?? {}), accept: "text/event-stream" }
  const response = await rawRequest(
    "POST",
    `/agents/stream/${encodeAgent(agent)}`,
    body,
    { ...http, headers },
  )
  if (response.body === null) return
  yield* parseSse(response.body)
}

/** Fetch the current execution status. */
export async function status(
  executionId: string,
  opts: HttpOptions = {},
): Promise<ExecutionStatus> {
  const raw = await request<unknown>(
    "GET",
    `/executions/${encodeURIComponent(executionId)}`,
    null,
    opts,
  )
  return executionStatusResponseSchema.parse(camelEnvelope(raw))
}

/** Cancel a running execution. */
export async function cancel(
  executionId: string,
  reason?: string,
  opts: HttpOptions = {},
): Promise<void> {
  // The REST surface accepts DELETE with an optional JSON body carrying
  // the reason — fall back to a query parameter on runtimes (Bun, certain
  // Workers) where DELETE bodies are stripped.
  const path = `/executions/${encodeURIComponent(executionId)}`
  await rawRequest("DELETE", path, reason !== undefined ? { reason } : null, opts)
}

/** HITL approval. */
export async function approve(
  executionId: string,
  opts: ApproveOptions = {},
): Promise<void> {
  const { byUserId, note, ...http } = opts
  const body: Record<string, unknown> = {}
  if (byUserId !== undefined) body.by_user_id = byUserId
  if (note !== undefined) body.note = note
  await rawRequest(
    "POST",
    `/executions/${encodeURIComponent(executionId)}/approve`,
    body,
    http,
  )
}

/** HITL denial. */
export async function deny(
  executionId: string,
  opts: DenyOptions = {},
): Promise<void> {
  const { byUserId, reason, ...http } = opts
  const body: Record<string, unknown> = {}
  if (byUserId !== undefined) body.by_user_id = byUserId
  if (reason !== undefined) body.reason = reason
  await rawRequest(
    "POST",
    `/executions/${encodeURIComponent(executionId)}/deny`,
    body,
    http,
  )
}

/** List executions awaiting human approval. */
export async function pendingApprovals(
  opts: HttpOptions = {},
): Promise<PendingApproval[]> {
  const raw = await request<unknown>("GET", "/approvals/pending", null, opts)
  // Apply camelEnvelope to each item to preserve user-defined payloads.
  const rawItems = (raw as { items?: unknown[] }).items ?? []
  const items = rawItems.map(camelEnvelope)
  const parsed = pendingApprovalsResponseSchema.parse({ items })
  return parsed.items
}

/** Namespace object — the shape `suite.agents` is built from. */
export const agents = {
  call,
  callAsync,
  stream,
  status,
  cancel,
  approve,
  deny,
  pendingApprovals,
} as const

File: packages/sdk-py/af_stack/agents.py
# SPDX-License-Identifier: Apache-2.0

"""``suite.agents.*`` — invoking AgentField from outside an agent.

Maps directly to the REST surface documented in ``TECH-SPEC.md §4.4``:

==================================  ====================================
SDK call                            Endpoint
==================================  ====================================
:func:`call`                        ``POST   /agents/{ns}.{func}``
:func:`call_async`                  ``POST   /agents/async/{ns}.{func}``
:func:`stream`                      ``GET    /agents/stream/{ns}.{func}``
:func:`status`                      ``GET    /executions/{id}``
:func:`cancel`                      ``DELETE /executions/{id}``
:func:`approve`                     ``POST   /executions/{id}/approve``
:func:`deny`                        ``POST   /executions/{id}/deny``
:func:`pending_approvals`           ``GET    /approvals/pending``
==================================  ====================================

This module deliberately does **not** expose a ``llm.chat`` style bypass.
All model calls flow through AgentField via these endpoints.
"""

from __future__ import annotations

from collections.abc import AsyncIterator
from typing import Any

from pydantic import BaseModel, ConfigDict, Field

from . import _http


class CallResult(BaseModel):
    """Synchronous ``agents.call`` response."""

    model_config = ConfigDict(extra="allow")

    execution_id: str = Field(..., description="AF execution id")
    output: Any = Field(default=None, description="Agent return value")
    status: str = Field(default="done")


class AsyncHandle(BaseModel):
    """Reference to an async execution started via ``agents.call_async``."""

    model_config = ConfigDict(extra="allow")

    execution_id: str
    status: str = "queued"
    webhook_url: str | None = None

    # Allow ``handle.id`` for ergonomic parity with TS clients.
    @property
    def id(self) -> str:
        return self.execution_id


class ExecutionStatus(BaseModel):
    """``agents.status`` response."""

    model_config = ConfigDict(extra="allow")

    execution_id: str
    status: str
    started_at: str | None = None
    ended_at: str | None = None
    output: Any = None
    error: dict[str, Any] | None = None


class PendingApproval(BaseModel):
    """One row from ``agents.pending_approvals``."""

    model_config = ConfigDict(extra="allow")

    execution_id: str
    agent: str
    requested_at: str | None = None
    reason: str | None = None
    metadata: dict[str, Any] = Field(default_factory=dict)


def _agent_path(name: str) -> str:
    """Validate ``ns.func`` shape and return the URL path segment."""
    if not name or "." not in name:
        raise ValueError(f"agent name must be of the form 'namespace.function', got {name!r}")
    ns, _, func = name.partition(".")
    if not ns or not func:
        raise ValueError(f"agent name must be of the form 'namespace.function', got {name!r}")
    return name


async def call(
    name: str,
    input: dict[str, Any] | None = None,
    *,
    timeout_s: float | None = None,
    metadata: dict[str, Any] | None = None,
) -> CallResult:
    """Synchronously call an AgentField agent and return its result.

    ``name`` must be of the form ``"namespace.function"``.
    """
    path = f"/agents/{_agent_path(name)}"
    payload: dict[str, Any] = {"input": input or {}}
    if metadata is not None:
        payload["metadata"] = metadata
    if timeout_s is not None:
        payload["timeout_s"] = timeout_s
    body = await _http.request_json(
        "POST",
        path,
        json=payload,
        timeout=timeout_s,
    )
    return CallResult.model_validate(body or {})


async def call_async(
    name: str,
    input: dict[str, Any] | None = None,
    *,
    webhook_url: str | None = None,
    metadata: dict[str, Any] | None = None,
) -> AsyncHandle:
    """Enqueue an agent execution and return immediately with a handle."""
    path = f"/agents/async/{_agent_path(name)}"
    payload: dict[str, Any] = {"input": input or {}}
    if webhook_url is not None:
        payload["webhook_url"] = webhook_url
    if metadata is not None:
        payload["metadata"] = metadata
    body = await _http.request_json("POST", path, json=payload)
    return AsyncHandle.model_validate(body or {})


async def stream(
    name: str,
    input: dict[str, Any] | None = None,
) -> AsyncIterator[dict[str, Any]]:
    """Stream events for an agent invocation via Server-Sent Events.

    Yields dicts of the shape ``{"event": str, "data": Any}``.
    """
    # TODO: end-to-end SSE test coverage. Currently exercised only via
    # unit tests on ``_http.stream_sse`` parser behaviour.
    path = f"/agents/stream/{_agent_path(name)}"
    payload: dict[str, Any] = {"input": input or {}}
    async for event in _http.stream_sse("POST", path, json=payload):
        yield event


async def status(execution_id: str) -> ExecutionStatus:
    """Fetch the current status of an execution."""
    body = await _http.request_json("GET", f"/executions/{execution_id}")
    return ExecutionStatus.model_validate(body or {})


async def cancel(execution_id: str, *, reason: str | None = None) -> None:
    """Cancel a running execution."""
    params = {"reason": reason} if reason else None
    await _http.request_json("DELETE", f"/executions/{execution_id}", params=params)


async def approve(
    execution_id: str,
    *,
    by_user_id: str | None = None,
    note: str | None = None,
) -> None:
    """Approve a paused HITL execution."""
    payload: dict[str, Any] = {}
    if by_user_id is not None:
        payload["by_user_id"] = by_user_id
    if note is not None:
        payload["note"] = note
    await _http.request_json("POST", f"/executions/{execution_id}/approve", json=payload)


async def deny(
    execution_id: str,
    *,
    reason: str | None = None,
    by_user_id: str | None = None,
) -> None:
    """Deny a paused HITL execution."""
    payload: dict[str, Any] = {}
    if reason is not None:
        payload["reason"] = reason
    if by_user_id is not None:
        payload["by_user_id"] = by_user_id
    await _http.request_json("POST", f"/executions/{execution_id}/deny", json=payload)


async def pending_approvals() -> list[PendingApproval]:
    """List executions currently waiting on a HITL approval."""
    body = await _http.request_json("GET", "/approvals/pending")
    rows = body if isinstance(body, list) else (body or {}).get("items") or []
    return [PendingApproval.model_validate(row) for row in rows]


__all__ = [
    "AsyncHandle",
    "CallResult",
    "ExecutionStatus",
    "PendingApproval",
    "approve",
    "call",
    "call_async",
    "cancel",
    "deny",
    "pending_approvals",
    "status",
    "stream",
]
